coreneuron/nrniv/main1.cpp

```diff
@@ -54,9 +55,78 @@ THE POSSIBILITY OF SUCH DAMAGE.
 #include "coreneuron/nrniv/partrans.h"
 #include "coreneuron/nrniv/multisend.h"
 #include "coreneuron/utils/file_utils.h"
+#include "coreneuron/nrniv/nrn2core_direct.h"
 #include <string.h>
 #include <climits>
 
+extern "C" {
+const char* corenrn_version() {
+    return coreneuron::bbcore_write_version;
+}
+
+int corenrn_embedded_run(int nthread, int have_gaps, int use_mpi, const char* arg) {
+    corenrn_embedded = 1;
+    corenrn_embedded_nthread = nthread;
+    coreneuron::nrn_have_gaps = have_gaps;
+
+#if defined(_OPENMP)
+    // if "export OMP_NUM_THREADS=nnn" is not set then omp by default sets
+    // the number of threads equal to the number of cores on this node.
+    // If there are a number of mpi processes on this node as well, things
+    // can go very slowly as there are so many more threads than cores.
+    // Assume the NEURON users pc.nthread() is well chosen if
+    // OMP_NUM_THREADS is not set.
+    if (!getenv("OMP_NUM_THREADS")) {
+        omp_set_num_threads(nthread);
+    }
+#endif
+
+    // count arg
+    int argc = 0;
+    int inarg = 0;
+    for (int i = 0; arg[i]; ++i) {
+        int sp = isspace(arg[i]);
+        if (inarg && sp) {  // start whitespace following previous arg
+            inarg = 0;
+        } else if (!inarg && !sp) {  // first char of current arg
+            argc++;
+            inarg = 1;
+        }
+    }
+    argc += use_mpi ? 2 : 1;  // corenrn -mpi or just corenrn
+    char** argv = new char*[argc];
+
+    // recount and fill argv
+    argc = 0;
+    argv[argc++] = strdup("corenrn");
+    if (use_mpi) {
+        argv[argc++] = strdup("-mpi");
```

Comment:
corresponding free missing?

Reply:
Yes.  In general, though, on return from nrncore_run it is not clear at the moment how much access NEURON needs to the coreneuron results. .eg. perhaps there should be an inverse
to nrnbbcore_write. Or, as we mentioned in the paper a series of accessor functions of various grain.